encrypt/ipsec_rotate_key.go

```diff
@@ -0,0 +1,183 @@
+	key        string
+	size       int
+	cipherMode string
+	cipherKey  string
+}
+
+// IPsecRotateKey rotates IPsec key.
+func (s *Encrypt) IPsecRotateKey(ctx context.Context) error {
+	ctx, cancelFn := context.WithTimeout(ctx, s.params.WaitDuration)
+	defer cancelFn()
+
+	secret, err := s.client.GetSecret(ctx, s.params.CiliumNamespace, defaults.EncryptionSecretName, metav1.GetOptions{})
+	if err != nil {
+		return fmt.Errorf("failed to fetch IPsec secret: %s", err)
+	}
+
+	key, err := ipsecKeyFromString(string(secret.Data["keys"]))
+	if err != nil {
+		return err
+	}
+
+	newKey, err := key.rotate()
+	if err != nil {
+		return fmt.Errorf("failed to rotate IPsec key: %s", err)
+	}
+
+	patch := []byte(`{"stringData":{"keys":"` + newKey.String() + `"}}`)
+	_, err = s.client.PatchSecret(ctx, s.params.CiliumNamespace, defaults.EncryptionSecretName, types.StrategicMergePatchType, patch, metav1.PatchOptions{})
+	if err != nil {
+		return fmt.Errorf("failed to patch IPsec secret with new key: %s", err)
+	}
+
+	_, err = fmt.Printf("IPsec key successfully rotated, new key SPI: %d\n", newKey.spi)
+	return err
+}
+
+func (k ipsecKey) String() string {
+	spiSuffix := ""
+	if k.spiSuffix {
+		spiSuffix = "+"
+	}
+	if k.cipherMode == "" {
+		return fmt.Sprintf("%d%s %s %s %d", k.spi, spiSuffix, k.algo, k.key, k.size)
+	}
+	return fmt.Sprintf("%d%s %s %s %s %s", k.spi, spiSuffix, k.algo, k.key, k.cipherMode, k.cipherKey)
+}
+
+var (
+	ipsecKeyRegex    = regexp.MustCompile(`^([[:digit:]]+\+?)[[:space:]](\S+)[[:space:]]([[:alnum:]]+)[[:space:]]([[:digit:]]+)$`)
+	cbcIPsecKeyRegex = regexp.MustCompile(`^([[:digit:]]+\+?)[[:space:]](\S+)[[:space:]]([[:alnum:]]+)[[:space:]](\S+)[[:space:]]([[:alnum:]]+)$`)
+	ipsecParsers     = map[*regexp.Regexp]func([]string) (ipsecKey, error){
+		ipsecKeyRegex:    keyFromSlice,
+		cbcIPsecKeyRegex: cbcKeyFromSlice,
+	}
+)
+
+func ipsecKeyFromString(s string) (ipsecKey, error) {
+	for matcher, parser := range ipsecParsers {
+		if matcher.MatchString(s) {
+			return parser(matcher.FindStringSubmatch(s))
+		}
+	}
+	return ipsecKey{}, fmt.Errorf("IPsec key has unsupported format")
+}
+
+func keyFromSlice(parts []string) (ipsecKey, error) {
+	if len(parts) != 5 {
+		return ipsecKey{}, fmt.Errorf("IPsec key invalid [expected parts: 5, actual parts: %d]", len(parts))
+	}
+	spiSuffix := false
+	if strings.HasSuffix(parts[1], "+") {
+		spiSuffix = true
+		parts[1] = strings.TrimSuffix(parts[1], "+")
+	}
+	spi, err := strconv.Atoi(parts[1])
+	if err != nil {
+		return ipsecKey{}, fmt.Errorf("invalid IPsec key SPI: %s", parts[1])
+	}
+	size, err := strconv.Atoi(parts[4])
+	if err != nil {
+		return ipsecKey{}, fmt.Errorf("invalid IPsec key size: %s", parts[4])
+	}
+	key := ipsecKey{
+		spi:       spi,
+		spiSuffix: spiSuffix,
+		algo:      parts[2],
+		key:       parts[3],
+		size:      size,
+	}
+	return key, nil
+}
+
+func cbcKeyFromSlice(parts []string) (ipsecKey, error) {
```

Comment:
This function name (and other instances of "`cbc`") should be changed to `cipherKeyFromSlice` or similar.

Reply:
Fixed.